pkg/querier/frontend/frontend_querier_queues.go

```diff
@@ -0,0 +1,216 @@
+		maxQueriers = 0
+	}
+
+	uq := q.userQueues[userID]
+
+	if uq == nil {
+		uq = &userQueue{
+			ch:    make(chan *request, q.maxUserQueueSize),
+			seed:  getSeedForUser(userID),
+			index: -1,
+		}
+		q.userQueues[userID] = uq
+
+		// Add user to the list of users... find first free spot, and put it there.
+		for ix, u := range q.users {
+			if u == "" {
+				uq.index = ix
+				q.users[ix] = userID
+				break
+			}
+		}
+
+		// ... or add to the end.
+		if uq.index < 0 {
+			uq.index = len(q.users)
+			q.users = append(q.users, userID)
+		}
+	}
+
+	if uq.maxQueriers != maxQueriers {
+		uq.maxQueriers = maxQueriers
+		uq.queriers = q.selectQueriersForUser(uq.seed, q.sortedQueriers, maxQueriers)
+	}
+
+	return uq.ch
+}
+
+// Finds next queue for the querier. To support fair scheduling between users, client is expected
+// to pass last user index returned by this function as argument. Is there was no previous
+// last user index, use -1.
+func (q *queues) getNextQueueForQuerier(lastUserIndex int, querier string) (chan *request, string, int) {
+	uid := lastUserIndex
+
+	for iters := 0; iters < len(q.users); iters++ {
+		uid = uid + 1
+
+		// Don't use "mod len(q.users)", as that could skip users at the beginning of the list
+		// for example when q.users has shrunk since last call.
+		if uid >= len(q.users) {
+			uid = 0
+		}
+
+		u := q.users[uid]
+		if u == "" {
+			continue
+		}
+
+		q := q.userQueues[u]
+
+		if q.queriers != nil {
+			if _, ok := q.queriers[querier]; !ok {
+				// This querier is not handling the user.
+				continue
+			}
+		}
+
+		return q.ch, u, uid
+	}
+	return nil, "", uid
+}
+
+func (q *queues) addQuerierConnection(querier string) {
+	conns := q.querierConnections[querier]
+
+	q.querierConnections[querier] = conns + 1
+
+	// First connection from this querier.
+	if conns == 0 {
+		q.sortedQueriers = append(q.sortedQueriers, querier)
+		sort.Strings(q.sortedQueriers)
+
+		q.recomputeUserQueriers()
+	}
+}
+
+func (q *queues) removeQuerierConnection(querier string) {
+	conns := q.querierConnections[querier]
+	if conns <= 0 {
+		return
+	}
+
+	conns--
+	if conns > 0 {
+		q.querierConnections[querier] = conns
+	} else {
+		delete(q.querierConnections, querier)
+
+		ix := sort.SearchStrings(q.sortedQueriers, querier)
+		q.sortedQueriers = append(q.sortedQueriers[:ix], q.sortedQueriers[ix+1:]...)
+
+		q.recomputeUserQueriers()
+	}
+}
+
+func (q *queues) recomputeUserQueriers() {
+	for _, uq := range q.userQueues {
+		uq.queriers = q.selectQueriersForUser(uq.seed, q.sortedQueriers, uq.maxQueriers)
+	}
+}
+
+func (q *queues) selectQueriersForUser(userSeed int64, allSortedQueriers []string, maxQueriers int) map[string]struct{} {
+	if maxQueriers == 0 || len(allSortedQueriers) <= maxQueriers {
+		// All queriers can be used.
+		return nil
+	}
+
+	result := make(map[string]struct{}, maxQueriers)
+	rnd := rand.New(rand.NewSource(userSeed))
+
+	for len(result) < maxQueriers {
+		ix := rnd.Intn(len(allSortedQueriers))
```

Comment:
this could roll the same querier multiple times?

Reply:
Yes it can. Alternative would be to make a copy of queriers, shuffle them and use first `maxQueriers` of them. Current version avoids this extra allocation at the cost of extra iterations. Not sure which one is better (benchmark time?)

Reply:
I've tried to use shuffling instead of trying random numbers. Updated benchmark shows little less time needed for `BenchmarkQueueRequest` (well, with +-7%, maybe not), but also little more memory allocations.

```
name              old time/op    new time/op    delta
GetNextRequest-4    59.7µs ± 4%    59.7µs ± 1%    ~     (p=0.645 n=10+9)
QueueRequest-4       626µs ± 7%     609µs ± 2%    ~     (p=0.105 n=10+10)

name              old alloc/op   new alloc/op   delta
GetNextRequest-4    1.60kB ± 0%    1.60kB ± 0%    ~     (all equal)
QueueRequest-4       322kB ± 0%     326kB ± 0%  +1.24%  (p=0.000 n=10+10)

name              old allocs/op  new allocs/op  delta
GetNextRequest-4       100 ± 0%       100 ± 0%    ~     (all equal)
QueueRequest-4       1.07k ± 0%     1.12k ± 0%  +4.68%  (p=0.000 n=10+10)
```

Reply:
I'm not sure if I understand this logic. Don't we want to shuffle-shard each user onto n distinct queriers? Does this function n distinct queriers?

Reply:
"Distinctness" is guaranteed by 1) using a map, 2) condition in the for loop. This loop doesn't end until there are `maxQueriers` distinct queriers in the map. In addition to that, thanks to `len(allSortedQueriers) > maxQueriers` precondition checked at the beginning of this method, we know that loop will eventually finish.

Reply:
Doing this via repeated random selections versus  shuffling the array seems like a probably unnecessary optimization at the expense of having a more deterministic outcome in terms of iterations required. I'm not sure it will matter practically, but seems like predictable would be better, meaning shuffle a copy of the array.

You don't have to shuffle the whole array, you can make a copy then from cnt over 0 to N first indices generate a random index between cnt+1 to N-1 inclisive and swap the value in the counter index with the random index so you can't ever select a duplicate, and you don't have to swap more values than the number you need to sellect. This is basically this: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm except I described it from 0 up and they describe it from N-1 down.

Reply:
I've changed the implementation to use suggested algorithm. (Btw, it's the same what `rand.Shuffle()` does, but in our case we can stop early, and set elements in returned map without another iteration).